Standardize Keypress Menus

Trenly opened the pull request:
@OfficialEsco @vedantmgoyal2009
I noticed most of the menus were similar, so I made a function to help reduce duplicated code. Please review

![image](https://user-images.githubusercontent.com/12611259/128909314-e7f2ceef-8ef5-47a2-997e-111a6a1b18f0.png)
![image](https://user-images.githubusercontent.com/12611259/128909377-6ce3a9d3-892f-45b0-b3fd-7329c677a735.png)
![image](https://user-images.githubusercontent.com/12611259/128920217-06cbe8a3-b67f-42c1-a4c8-661ec05f5a0a.png)

OfficialEsco commented:
Did https://github.com/Trenly/winget-pkgs/pull/11/commits/9ef559c5b808497de68f05e6fbe93d676c9f04c4 fix this? 😁

```
Have you signed the Contributor License Agreement (CLA)?
Reference Link: https://cla.opensource.microsoft.com/microsoft/winget-pkgs
  [Y] Yes
  [N] No

Enter Choice (default is 'N'):

Have you checked that there aren't other open pull requests for the same manifest update/change?
Reference Link: https://github.com/microsoft/winget-pkgs/pulls
  [Y] Yes
  [N] No

Enter Choice (default is 'N'): yMethod invocation failed because [System.ConsoleKey] does not contain a method named 'Replace'.
At C:\Users\Esco\GitHub\winget-pkgs\Tools\YamlCreate.ps1:888 char:27
+ ...         'Y' { $PrBodyContentReply += $_.Replace("[ ]", "[X]"), "`n" }
+                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : MethodNotFound
 

Automatic manifest validation failed. Check your manifest and try again


Have you validated your manifest locally with 'winget validate --manifest <path>'

  [Y] Yes
  [N] No

Enter Choice (default is 'N'): yMethod invocation failed because [System.ConsoleKey] does not contain a method named 'Replace'.
At C:\Users\Esco\GitHub\winget-pkgs\Tools\YamlCreate.ps1:907 char:31
+ ...         'Y' { $PrBodyContentReply += $_.Replace("[ ]", "[X]"), "`n" }
+                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : MethodNotFound
 

You did not test your Manifest in Windows Sandbox previously.


Have you tested your manifest locally with 'winget install --manifest <path>'

  [Y] Yes
  [N] No

Enter Choice (default is 'N'): n

Does your manifest conform to the 1.0 schema?
Reference Link: https://github.com/microsoft/winget-cli/blob/master/doc/ManifestSpecv1.0.md
  [Y] Yes
  [N] No

Enter Choice (default is 'N'): yMethod invocation failed because [System.ConsoleKey] does not contain a method named 'Replace'.
At C:\Users\Esco\GitHub\winget-pkgs\Tools\YamlCreate.ps1:942 char:27
+ ...         'Y' { $PrBodyContentReply += $_.Replace("[ ]", "[X]"), "`n" }
+                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : MethodNotFound
 


Does this pull request resolve any issues?

  [Y] Yes
  [N] No

Enter Choice (default is 'N'): n
Method invocation failed because [System.ConsoleKey] does not contain a method named 'Trim'.
At C:\Users\Esco\GitHub\winget-pkgs\Tools\YamlCreate.ps1:980 char:5
+     $PrBodyContentReply = ($PrBodyContentReply.Trim() -ne '')
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : MethodNotFound
 
```

Output =
```
Backspace

N
```

Trenly commented:
> Did [9ef559c](https://github.com/Trenly/winget-pkgs/commit/9ef559c5b808497de68f05e6fbe93d676c9f04c4) fix this? 😁
> 
> ```
> Have you signed the Contributor License Agreement (CLA)?
> Reference Link: https://cla.opensource.microsoft.com/microsoft/winget-pkgs
>   [Y] Yes
>   [N] No
> 
> Enter Choice (default is 'N'):
> 
> Have you checked that there aren't other open pull requests for the same manifest update/change?
> Reference Link: https://github.com/microsoft/winget-pkgs/pulls
>   [Y] Yes
>   [N] No
> 
> Enter Choice (default is 'N'): yMethod invocation failed because [System.ConsoleKey] does not contain a method named 'Replace'.
> At C:\Users\Esco\GitHub\winget-pkgs\Tools\YamlCreate.ps1:888 char:27
> + ...         'Y' { $PrBodyContentReply += $_.Replace("[ ]", "[X]"), "`n" }
> +                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
>     + FullyQualifiedErrorId : MethodNotFound
>  
> 
> Automatic manifest validation failed. Check your manifest and try again
> 
> 
> Have you validated your manifest locally with 'winget validate --manifest <path>'
> 
>   [Y] Yes
>   [N] No
> 
> Enter Choice (default is 'N'): yMethod invocation failed because [System.ConsoleKey] does not contain a method named 'Replace'.
> At C:\Users\Esco\GitHub\winget-pkgs\Tools\YamlCreate.ps1:907 char:31
> + ...         'Y' { $PrBodyContentReply += $_.Replace("[ ]", "[X]"), "`n" }
> +                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
>     + FullyQualifiedErrorId : MethodNotFound
>  
> 
> You did not test your Manifest in Windows Sandbox previously.
> 
> 
> Have you tested your manifest locally with 'winget install --manifest <path>'
> 
>   [Y] Yes
>   [N] No
> 
> Enter Choice (default is 'N'): n
> 
> Does your manifest conform to the 1.0 schema?
> Reference Link: https://github.com/microsoft/winget-cli/blob/master/doc/ManifestSpecv1.0.md
>   [Y] Yes
>   [N] No
> 
> Enter Choice (default is 'N'): yMethod invocation failed because [System.ConsoleKey] does not contain a method named 'Replace'.
> At C:\Users\Esco\GitHub\winget-pkgs\Tools\YamlCreate.ps1:942 char:27
> + ...         'Y' { $PrBodyContentReply += $_.Replace("[ ]", "[X]"), "`n" }
> +                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
>     + FullyQualifiedErrorId : MethodNotFound
>  
> 
> 
> Does this pull request resolve any issues?
> 
>   [Y] Yes
>   [N] No
> 
> Enter Choice (default is 'N'): n
> Method invocation failed because [System.ConsoleKey] does not contain a method named 'Trim'.
> At C:\Users\Esco\GitHub\winget-pkgs\Tools\YamlCreate.ps1:980 char:5
> +     $PrBodyContentReply = ($PrBodyContentReply.Trim() -ne '')
> +     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
>     + FullyQualifiedErrorId : MethodNotFound
>  
> ```
> 
> Output =
> 
> ```
> Backspace
> 
> N
> ```

I believe it did

OfficialEsco commented:
Yapyap that did, haven't tested the alternative routes tho, but i'm guessing this fixed every similar string

Trenly commented:
> Yapyap that did, haven't tested the alternative routes tho, but i'm guessing this fixed every similar string

I think it did. I’m not sure why the pipeline output was the iterand and not a variable since it can really mess things up, but yeah, should be fixed

OfficialEsco commented:
Seems okay, just a bit weird in the start compared to the last 2 version's we have had of the prompts 😁
Also just noticed mine is different compared to the screenshots, when the Help Variable it empty it outputs a space
```
Have you tested your manifest locally with 'winget install --manifest <path>'

  [Y] Yes
  [N] No

Enter Choice (default is 'N'):
```

Can you add a quick fix to this PR?
Fixing this spacing

```
Do you want to submit your PR now?

  [Y] Yes
  [N] No

Enter Choice (default is 'Y'): yHEAD is now at 78f8884e5 New version: Telegram.TelegramDesktop version 2.9.2 (#24434)
warning: LF will be replaced by CRLF in manifests/j/Joplin/Joplin/2.2.7/Joplin.Joplin.installer.yaml.
```

This also happens when you accept to test in Sandbox, those are the 2 places i remember atm

Just noticed another issue when i said mine looked different, this prompt is messed up
```
You did not test your Manifest in Windows Sandbox previously.


Have you tested your manifest locally with 'winget install --manifest <path>'

  [Y] Yes
  [N] No

Enter Choice (default is 'N'):
```

OfficialEsco commented on Tools/YamlCreate.ps1:
```diff
@@ -835,163 +860,137 @@ Function Test-Manifest {
-                        $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
-                    } else {
-                        $PrBodyContentReply += $_, "`n"
+                    Write-Host -ForegroundColor 'Blue' "You did not test your Manifest in Windows Sandbox previously."
+                    $_menu = @{
+                        entries       = @("[Y] Yes"; "*[N] No")
+                        Prompt        = "Have you tested your manifest locally with 'winget install --manifest <path>'"
+                        DefaultString = "N"
+                    }
```
This is probably okay

```suggestion
                    $_menu = @{
                        entries       = @("[Y] Yes"; "*[N] No")
                        Prompt        = "You did not test your Manifest in Windows Sandbox previously."
                        HelpText      = "Have you tested your manifest locally with 'winget install --manifest <path>'?"
                        DefaultString = "N"
                    }
```

OfficialEsco commented on Tools/YamlCreate.ps1:
```diff
@@ -835,163 +860,137 @@ Function Test-Manifest {
+                }
+                else {
                     Write-Host
                     Write-Host -ForegroundColor 'Red' "Automatic manifest validation failed. Check your manifest and try again"
```
```suggestion
```

OfficialEsco commented on Tools/YamlCreate.ps1:
```diff
@@ -835,163 +860,137 @@ Function Test-Manifest {
-                        $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
-                    } else {
-                        $PrBodyContentReply += $_, "`n"
+                    $_menu = @{
+                        entries       = @("[Y] Yes"; "*[N] No")
+                        Prompt        = "Have you validated your manifest locally with 'winget validate --manifest <path>'"
+                        DefaultString = "N"
+                    }
```
```suggestion
                    $_menu = @{
                        entries       = @("[Y] Yes"; "*[N] No")
                        Prompt        = "Automatic manifest validation failed. Check your manifest and try again"
                        HelpText      = "Have you validated your manifest locally with 'winget validate --manifest <path>'?"
                        DefaultString = "N"
                    }
```

OfficialEsco commented on Tools/YamlCreate.ps1:
```diff
@@ -835,163 +860,137 @@ Function Test-Manifest {
+                    $PrBodyContentReply += $_line.Replace("[ ]", "[X]"), "`n"
+                }
+                else {
                     Write-Host
```
Is this needed?
```suggestion
```

OfficialEsco commented on Tools/YamlCreate.ps1:
```diff
@@ -835,163 +860,137 @@ Function Test-Manifest {
+                    $PrBodyContentReply += $_line.Replace("[ ]", "[X]"), "`n"
+                }
+                else {
                     Write-Host
```
Is this needed?
```suggestion
```

Trenly commented:
I'll look at this and make a single commit for your suggestions; Do you think it would be better to have the error messages as the prompt like your suggestion . . . 
```powershell
$_menu = @{
    entries       = @("[Y] Yes"; "*[N] No")
    Prompt        = "Automatic manifest validation failed. Check your manifest and try again"
    HelpText      = "Have you validated your manifest locally with 'winget validate --manifest <path>'?"
    DefaultString = "N"
 }
```
or as the help text, since it isn't the actual question being asked like this -
```powershell
$_menu = @{
    entries       = @("[Y] Yes"; "*[N] No")
    Prompt        = "Have you validated your manifest locally with 'winget validate --manifest <path>'?"
    HelpText      = "Automatic manifest validation failed. Check your manifest and try again"
    DefaultString = "N"
 }
```

OfficialEsco commented:
> I'll look at this and make a single commit for your suggestions; Do you think it would be better to have the error messages as the prompt like your suggestion . . .

> or as the help text, since it isn't the actual question being asked like this -

Yeye thats fine, i think it should be at the top with maybe red text (something noticeable) if possible, i just worked with what i had 😁 Which.. failed
```
Have you checked that there aren't other open pull requests for the same manifest update/change?
Reference Link: https://github.com/microsoft/winget-pkgs/pulls
  [Y] Yes
  [N] No

Enter Choice (default is 'N'): y

You did not test your Manifest in Windows Sandbox previously.

  [Y] Yes
  [N] No

Enter Choice (default is 'N'):
```

OfficialEsco commented:
Lets leave it like that for now, can you add do this quick change for cleanness sake
```
        git fetch upstream master -q
        git switch -d upstream/master
        if ($LASTEXITCODE -eq '0') {
            git add -A
            git commit -m "$CommitType`: $PackageIdentifier version $PackageVersion"
            git switch -c "$PackageIdentifier-$PackageVersion"
            git push --set-upstream origin "$PackageIdentifier-$PackageVersion"
```
to
```
        git fetch upstream master --quiet
        git switch -d upstream/master
        if ($LASTEXITCODE -eq '0') {
            git add -A
            git commit -m "$CommitType`: $PackageIdentifier version $PackageVersion" --quiet

            git switch -c "$PackageIdentifier-$PackageVersion" --quiet
            git push --set-upstream origin "$PackageIdentifier-$PackageVersion" --quiet
```

need the first `git switch` to not be silent to get this line `HEAD is now at b73a14ccd gitthermal.thermal version 0.0.4 (#24445)`

```
warning: LF will be replaced by CRLF in manifests/s/SoftwareFreedomConservancy/QEMU/6.0.93/SoftwareFreedomConservancy.QEMU.yaml.
The file will have its original line endings in your working directory
```
Is probably from `git add` which cannot be silenced

```
remote: 
remote: Create a pull request for 'SoftwareFreedomConservancy.QEMU-6.0.93' on GitHub by visiting:
remote:      https://github.com/OfficialEsco/winget-pkgs/pull/new/SoftwareFreedomConservancy.QEMU-6.0.93
remote:
```
is believe is from `gh pr create` which does not have a silent switch

It looks a lot better tho

OfficialEsco commented:
Think i've found a new issue which might have appeared after the -exclude problem popped up

```
[Required] Enter the Package Identifier, in the following format <Publisher shortname.Application shortname>. For example: Microsoft.Excel
PackageIdentifier: TorProject.TorBrowser

[Required] Enter the version. for example: 1.33.7
Version: 10.5.4
Found Existing Version: 10.5a17
Get-ChildItem : Cannot find path 'C:\Users\Esco\GitHub\winget-pkgs\manifests\t\TorProject\TorBrowser\10.5a17' because it does not 
exist.
At C:\Users\Esco\GitHub\winget-pkgs\Tools\YamlCreate.ps1:1286 char:36
+ ... cript:OldManifests = Get-ChildItem -Path "$AppFolder\..\$LastVersion"
+                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (C:\Users\Esco\...Browser\10.5a17:String) [Get-ChildItem], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.GetChildItemCommand
 
Error: Version 10.5a17 does not contain the required manifests
At C:\Users\Esco\GitHub\winget-pkgs\Tools\YamlCreate.ps1:1305 char:9      
+         Throw "Error: Version $LastVersion does not contain the requi ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    
    + CategoryInfo          : OperationStopped: (Error: Version ...uired manifests:String) [], RuntimeException
    + FullyQualifiedErrorId : Error: Version 10.5a17 does not contain the required manifests
```

It picked up `\Manifests\t\TorProject\TorBrowser\Alpha\10.5a17\` instead of `\Manifests\t\TorProject\TorBrowser\10.5.2`

Trenly commented:
> Think i've found a new issue which might have appeared after the -exclude problem popped up
> 
> ```
> [Required] Enter the Package Identifier, in the following format <Publisher shortname.Application shortname>. For example: Microsoft.Excel
> PackageIdentifier: TorProject.TorBrowser
> 
> [Required] Enter the version. for example: 1.33.7
> Version: 10.5.4
> Found Existing Version: 10.5a17
> Get-ChildItem : Cannot find path 'C:\Users\Esco\GitHub\winget-pkgs\manifests\t\TorProject\TorBrowser\10.5a17' because it does not 
> exist.
> At C:\Users\Esco\GitHub\winget-pkgs\Tools\YamlCreate.ps1:1286 char:36
> + ... cript:OldManifests = Get-ChildItem -Path "$AppFolder\..\$LastVersion"
> +                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : ObjectNotFound: (C:\Users\Esco\...Browser\10.5a17:String) [Get-ChildItem], ItemNotFoundException
>     + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.GetChildItemCommand
>  
> Error: Version 10.5a17 does not contain the required manifests
> At C:\Users\Esco\GitHub\winget-pkgs\Tools\YamlCreate.ps1:1305 char:9      
> +         Throw "Error: Version $LastVersion does not contain the requi ...
> +         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    
>     + CategoryInfo          : OperationStopped: (Error: Version ...uired manifests:String) [], RuntimeException
>     + FullyQualifiedErrorId : Error: Version 10.5a17 does not contain the required manifests
> ```
> 
> It picked up `\Manifests\t\TorProject\TorBrowser\Alpha\10.5a17\` instead of `\Manifests\t\TorProject\TorBrowser\10.5.2`

Looks like exclude overrides the depth flag

OfficialEsco commented:
> Edit - Can we assume that all manifest files match the pattern `*.yaml` ?

Very much yes, i think i've actually thought about that before when i was looking trough the script, but didn't look much into it since it actually worked..

Trenly commented:
> > Edit - Can we assume that all manifest files match the pattern `*.yaml` ?
> 
> Very much yes, i think i've actually thought about that before when i was looking trough the script, but didn't look much into it since it actually worked..

Okay; I think I fixed the filter.

I also changed the git code to be what you requested, and also suppressed the crlf warnings

Trenly commented on Tools/YamlCreate.ps1:
```diff
@@ -1270,8 +1292,9 @@ Function Read-PreviousWinGet-Manifest-Yaml {
-            $script:ExistingVersions = Split-Path (Split-Path (Get-ChildItem -Path "$AppFolder\..\" -Recurse -Depth 1 -File -Exclude ".validation").FullName ) -Leaf | Sort-Object $ToNatural | Select-Object -Unique
+            $script:LastVersion = Split-Path (Split-Path (Get-ChildItem -Path "$AppFolder\..\" -Recurse -Depth 1 -File -Filter "*.yaml").FullName ) -Leaf | Sort-Object $ToNatural | Select-Object -Last 1
+            $script:ExistingVersions = Split-Path (Split-Path (Get-ChildItem -Path "$AppFolder\..\" -Recurse -Depth 1 -File -Filter "*.yaml").FullName ) -Leaf | Sort-Object $ToNatural | Select-Object -Unique
+            Write-Host Split-Path (Get-ChildItem -Path "$AppFolder\..\" -Recurse -Depth 1 -File -Filter "*.yaml").FullName 
```
Whoops

OfficialEsco commented:
Well... There appears to be a new issue where it dosn't detect the manifests in the sub folder >.<
Was trying to update Google.Chrome.Beta and it didn't fetch any metadata...

Trenly commented:
@OfficialEsco I see an email that you posted a comment; Did you delete the comment?

Trenly commented:
> Yes, i was silly and on the wrong branch after fixing a Manifest🤦‍♂️ Somehow the script in the official Repo (v1.1.5) maintain some of the functions from the new script so it looks identical, however it does not fetch the new method of fetching metadata, i've checked and realized before, but now i've added `;git switch YamlParsing` so i was 99% sure i was on the right branch... So i realized once i opened GitHub Desktop and about to switch to your branch 🤦‍♂️🤦‍♂️

All good! Just wanted to maks sure I wasn’t missing something
